typing/jkind.ml

```diff
@@ -1457,94 +1466,105 @@ module Jkind_desc = struct
+        let ti : With_bounds_type_info.t =
+          { relevant_axes =
+              Axis_set.diff ti.relevant_axes (Bounds.get_max_axes bounds_so_far)
+          }
         in
-        match Loop_control.check ctl ty with
-        | Stop ->
-          (* out of fuel *)
-          join_respecting_omit Bounds.max, With_bounds_types.of_list bs
-        | Skip -> loop ctl bounds_so_far bs (* skip [b] *)
-        | Continue ctl_after_unpacking_b -> (
-          match jkind_of_type ty with
-          | Some b_jkind ->
-            if (match b_jkind.quality with Not_best -> false | Best -> true)
-               || not require_best
-            then
+        match Axis_set.is_empty ti.relevant_axes with
+        | true -> loop ctl bounds_so_far bs
+        | false -> (
+          let join_bounds b1 b2 ~relevant_axes =
+            Bounds.Map2.f
+              { f =
+                  (fun (type a) ~(axis : a Axis.t) b1 b2 ->
+                    if Axis_set.mem relevant_axes axis
+                    then
+                      let (module Bound_ops) = Axis.get axis in
+                      Bound_ops.join b1 b2
+                    else b1)
+              }
+              b1 b2
+          in
+          let found_jkind_for_ty new_ctl b_upper_bounds b_with_bounds quality :
+              Bounds.t * (l2 * r2) with_bounds * Fuel_status.t =
+            match quality, mode with
+            | Best, _ | Not_best, Ignore_best ->
               let bounds_so_far =
-                join_respecting_omit b_jkind.jkind.mod_bounds
+                join_bounds bounds_so_far b_upper_bounds
+                  ~relevant_axes:ti.relevant_axes
               in
               (* Descend into the with-bounds of each of our with-bounds types'
                   with-bounds *)
-              let bounds_so_far, nested_with_bounds =
-                loop ctl_after_unpacking_b bounds_so_far
-                  (With_bounds.to_list b_jkind.jkind.with_bounds)
-              in
-              (* Use *original* ctl here, so we don't fall over on
-                  a record with 20 lists with different payloads. *)
-              let bounds, bs' = loop ctl_after_unpacking_b bounds_so_far bs in
-              bounds, With_bounds.join_bounds_maybe_empty nested_with_bounds bs'
-            else
-              (* Don't know the best kind for this type, so skip it, leaving it in the
-                 (normalized) with_bounds. *)
-              let bounds_so_far =
-                if require_best
-                then bounds_so_far
-                else join_respecting_omit Bounds.max
+              let bounds_so_far, nested_with_bounds, fuel_result1 =
+                loop new_ctl bounds_so_far (With_bounds.to_list b_with_bounds)
               in
-              let bounds_so_far, bs' =
-                loop ctl_after_unpacking_b bounds_so_far bs
+              (* CR: which ctl to use?
+                 (* Use *original* ctl here, so we don't fall over on
```

Comment:
Especially with the stress on `original`, it's a little confusing using `new_ctl` (the `new` bit). At this point in my read, I don't fully understand this yet so can't suggest a concrete improvement.

Reply:
The comment stating "original" is an old comment from when we really did use the original one. I think aspen's pr might also make this same change, so I think the discussion over which ctl to use should happen there.